### PR TITLE
Fix ICE when there is a non-Unicode entry in the incremental crate directory

### PR DESCRIPTION
**Pull request by beetrees**
Fix the ICE that occurs when there is a non-Unicode entry in the incremental crate directory by replacing uses of `to_string_lossy` + `assert_no_characters_lost` with `to_str`. The added test would cause the compiler to ICE before this PR.

### CODE REVIEW

**Comment by Nadrieril** [comment]
Thanks!

@bors r+ rollup

**Comment by workingjubilee** [comment]
I believe this is failing rollups:

https://github.com/rust-lang/rust/pull/124133#issuecomment-2065190547

@bors r-

**Comment by workingjubilee** [comment]
I tested my hypothesis and was wrong.

@bors r=Nadrieril

File: tests/run-make/non-unicode-in-incremental-dir/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+use run_make_support::{libc, rustc, tmp_dir};
+
+fn main() {
+    #[cfg(unix)]
+    let non_unicode: &std::ffi::OsStr = std::os::unix::ffi::OsStrExt::from_bytes(&[0xFF]);
+    #[cfg(windows)]
+    let non_unicode: std::ffi::OsString = std::os::windows::ffi::OsStringExt::from_wide(&[0xD800]);
+    match std::fs::create_dir(tmp_dir().join(&non_unicode)) {
+        Err(e) if e.raw_os_error() == Some(libc::EILSEQ) => {
```
Comment (ChrisDenton) [inline tests/run-make/non-unicode-in-incremental-dir/rmake.rs:11]: That error code assumes create_dir is implemented using a POSIX API. That isn't true on Windows and may not even be true on Unix OSes (depending on how the create_dir is implemented).

Comment (Nadrieril) [reply tests/run-make/non-unicode-in-incremental-dir/rmake.rs:11]: I don't see `EILSEQ` referenced anywhere in `std` so I don't think there's a portable way to detect this. You can always try to create a file with a unicode path first to ensure the test doesn't silently fail for other permissions reasons.

**Comment by Nadrieril** [comment]
Thanks!

@bors r+

**Comment by Nadrieril** [comment]
Looks unrelated

@bors retry

**Pull request merged**